blocks/bricks/bn.py

```diff
@@ -0,0 +1,345 @@
+
+    @property
+    def num_output_channels(self):
+        return self.num_channels
+
+
+class SpatialBatchNormalization(BatchNormalization):
+    """Convenient subclass for batch normalization across spatial inputs.
+
+    Parameters
+    ----------
+    input_dim : int or tuple
+        The input size of a single example. Must be length at least 2.
+        It's assumed that the first axis of this tuple is a "channels"
+        axis, which should not be summed over, and all remaining
+        dimensions are spatial dimensions.
+
+    Notes
+    -----
+    See :class:`BatchNormalization` for more details (and additional
+    keyword arguments).
+
+    """
+    @lazy(allocation=['input_dim'])
+    def __init__(self, input_dim, **kwargs):
+        if not isinstance(input_dim,
+                          collections.Sequence) or len(input_dim) < 2:
+            raise ValueError('expected input_dim to be length >= 2 '
+                             '(channels, height, width)')
+        broadcastable = (False,) + ((True,) * (len(input_dim) - 1))
+        kwargs.setdefault('broadcastable', broadcastable)
+        super(SpatialBatchNormalization, self).__init__(input_dim, **kwargs)
+
+
+class BatchNormalizedMLP(MLP):
+    """Convenient subclass for building an MLP with batch normalization.
+
+    Parameters
+    ----------
+    save_memory : bool, optional
+        See :class:`BatchNormalization`.
+
+    Notes
+    -----
+    All other parameters are the same as :class:`~blocks.bricks.MLP`. Each
+    activation brick is wrapped in a :class:`~blocks.bricks.Sequence`
+    containing an appropriate :class:`BatchNormalization` brick and
+    the activation that follows it.
+
+    By default, the contained :class:`~blocks.bricks.Linear` bricks will
+    not contain any biases, as they could be canceled out by the biases
+    in the :class:`BatchNormalization` bricks being added. Pass
+    `use_bias` with a value of `True` if you really want this for some
+    reason.
+
+    """
+    @lazy(allocation=['dims'])
+    def __init__(self, activations, dims, *args, **kwargs):
+        save_memory = kwargs.pop('save_memory', True)
+        activations = [
+            Sequence([BatchNormalization(save_memory=save_memory).apply,
+                      act.apply], name='batch_norm_activation_{}'.format(i))
+            for i, act in enumerate(activations)
+        ]
+        # Batch normalization bricks incorporate a bias, so there's no
+        # need for our Linear bricks to have them.
+        kwargs.setdefault('use_bias', False)
+        super(BatchNormalizedMLP, self).__init__(activations, dims, *args,
+                                                 **kwargs)
+
+    @property
+    def save_memory(self):
+        return self._save_memory
+
+    @save_memory.setter
+    def save_memory(self, value):
+        self._save_memory = value
+        for act in self.activations:
+            assert isinstance(act.children[0], BatchNormalization)
```

Comment:
This assertion would be nice to have in `_push_allocation_config` below as well.